kernel/sem.c

```diff
@@ -163,3 +189,36 @@ int k_sem_take(struct k_sem *sem, s32_t timeout)
 
 	return _Swap(key);
 }
+
+#ifdef CONFIG_USERSPACE
+u32_t _handler_k_sem_take(u32_t sem_ptr, u32_t timeout, u32_t arg3,
+			  u32_t arg4, u32_t arg5, u32_t arg6, void *ssf)
+{
+	_SYSCALL_ARG2;
+
+	_SYSCALL_IS_OBJ(sem_ptr, K_OBJ_SEM, 0, ssf);
```

Comment:
Just thinking out loud here: if we grouped syscalls by the object they operate on, maybe the validation could be performed in only one place, reducing slightly the code size and the syscall table.  The downside is that this would require more branches (or indirect calls), though.

Reply:
hmm. not sure I want the extra overhead associated with this. i think performance trumps ease of use in this case. once the initial batch of APIs is out I don't think we're going to be adding new system calls, or introducing new driver subsystems frequently enough to make it worth it.
right now, the system call assembly code jumps right into the handler function, very fast.